ZukoAzula2016/src/org/team1540/drivers/kauailabs/AHRSRegister.java

```diff
@@ -0,0 +1,197 @@
+
+    MAG_X(0x40, 1 / 0.15f, RegisterType.U16_MUL, WriteMode.READ_ONLY), MAG_Y(0x42, 1 / 0.15f, RegisterType.U16_MUL, WriteMode.READ_ONLY), MAG_Z(0x44, 1 / 0.15f, RegisterType.U16_MUL, WriteMode.READ_ONLY), PRESSURE_I(0x46, RegisterType.U16),
+
+    PRESSURE_D(0x48, RegisterType.SF32), PRESSURE_TEMP(0x4A, RegisterType.U16), YAW_OFFSET(0x4C, RegisterType.U16), QUAT_OFFSET_W(0x4E, RegisterType.U16),
+
+    QUAT_OFFSET_X(0x50, RegisterType.U16), QUAT_OFFSET_Y(0x52, RegisterType.U16), QUAT_OFFSET_Z(0x54, RegisterType.U16), INTEGRATION_CTL(0x56, RegisterType.U8, WriteMode.WRITE_ONLY), PAD_UNUSED(0x57, RegisterType.U8, WriteMode.WRITE_ONLY),
+
+    VEL_X_I(0x58, RegisterType.U16), VEL_X_D(0x5A, RegisterType.U16), VEL_Y_I(0x5C, RegisterType.U16), VEL_Y_D(0x5E, RegisterType.U16),
+
+    VEL_Z_I(0x60, RegisterType.U16), VEL_Z_D(0x62, RegisterType.U16), DISP_X_I(0x64, RegisterType.U16), DISP_X_D(0x66, RegisterType.U16),
+
+    DISP_Y_I(0x68, RegisterType.U16), DISP_Y_D(0x6A, RegisterType.U16), DISP_Z_I(0x6C, RegisterType.U16), DISP_Z_D(0x6E, RegisterType.U16);
+    
+    static enum RegisterType {
+        // L is followed by H
+        U8(1), U16(2), U16_MUL(2), U32(4), SF16_100(2), UF16_100(2), SF16_1000(2), SF32(4), FLAG16(2);
+
+        public final int bytes;
+
+        private RegisterType(int bytes) {
+            this.bytes = bytes;
+        }
+    }
+
+    static enum WriteMode {
+        READ_ONLY, WRITE_ONLY, READ_WRITE;
+
+        public boolean canRead() {
+            return this != WRITE_ONLY;
+        }
+
+        public boolean canWrite() {
+            return this != READ_ONLY;
+        }
+    }
+
+    final byte registerID;
+    final RegisterType type;
+    final WriteMode mode;
+    final double extra;
+
+    private AHRSRegister(int registerID, double extra, RegisterType rt, WriteMode mode) {
+        if (registerID < 0 || registerID >= 128 || rt == null || mode == null) {
+            throw new IllegalArgumentException();
+        }
+        if (extra != 0 && rt != RegisterType.FLAG16 && rt != RegisterType.U16_MUL) {
+            throw new IllegalArgumentException(); // must not pass extra
+        }
+        this.extra = extra;
+        this.registerID = (byte) registerID;
+        this.type = rt;
+        this.mode = mode;
+    }
+
+    private AHRSRegister(int registerID, RegisterType rt, WriteMode mode) {
+        this(registerID, 0, rt, mode);
+        if (rt == RegisterType.FLAG16) {
+            throw new IllegalArgumentException(); // must use previous
+                                                  // constructor
+        }
+    }
+
+    private AHRSRegister(int registerID, RegisterType rt) {
+        this(registerID, rt, WriteMode.READ_ONLY);
+    }
+
+    private AHRSRegister(int registerID) {
+        this(registerID, RegisterType.U8, WriteMode.READ_ONLY);
+    }
+
+    public byte decodeByte(RegisterIO io) {
```

Comment:
If this is supposed to be unsigned, maybe you want to have them all return a larger integer type?